% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/query_vcf.R
\name{vcf_query}
\alias{vcf_query}
\title{Query a VCF File For Numeric Genotypes}
\usage{
vcf_query(file, regions, samples, verbose = FALSE, recode = TRUE)
}
\arguments{
\item{file}{The name of the indexed vcf file.}

\item{regions}{A data frame containing chrom, start, and end of the regions to be extracted.}

\item{samples}{A character vector of sample names to be extracted.}

\item{verbose}{Whether to output with \code{cat()} the bcftools calls.}

\item{recode}{Whether to recode genotypes as numeric (assuming diallelic sites only).}
}
\value{
An integer matrix, giving the number of nonreference alleles for each sample.
For instance, "0/0", "0|0", and "0" are coded as 0, while "0/1", "2|0", or "1" are coded as 1,
and "1/1", "1/2" or "1|4" are all coded as 2.

The format is determined by looking at the first 100 loci in the first 100 individuals.

bcftools requires that vcf files are bgzipped, and both bcf and vcf files must be indexed,
so if you have a (plain text) vcf file "my_data.vcf", run:\cr
\cr  
  bgzip my_data.vcf\cr
  bcftools index my_data.vcf.gz\cr
\cr  
and then use file="my_data.vcf.gz".
}
\description{
Uses bcftools to query vcf or bcf files for genotypes in a collection of regions,
returning the genotypes in numeric format, giving the number of nonreference alleles.
Assumes the data are in the format either "0/1", "0|1" (for diploids), or "0" (for haploids).
Data are returned in the order encountered in the VCF file even if the regions are not in this order.
}
